crates/autopilot/src/circuit_breaker.rs

```diff
@@ -0,0 +1,208 @@
+//! This module is responsible for updating the database, for each settlement
+//! event that is emitted by the settlement contract.
+//
+// When we put settlement transactions on chain there is no reliable way to
+// know the transaction hash because we can create multiple transactions with
+// different gas prices. What we do know is the account and nonce that the
+// transaction will have which is enough to uniquely identify it.
+//
+// We build an association between account-nonce and tx hash by backfilling
+// settlement events with the account and nonce of their tx hash. This happens
+// in an always running background task.
+//
+// Alternatively we could change the event insertion code to do this but I (vk)
+// would like to keep that code as fast as possible to not slow down event
+// insertion which also needs to deal with reorgs. It is also nicer from a code
+// organization standpoint.
+
+// 2. Inserting settlement observations
+//
+// see database/sql/V048__create_settlement_rewards.sql
+//
+// Surplus and fees calculation is based on:
+// a) the mined transaction call data
+// b) the auction external prices fetched from orderbook
+// c) the orders fetched from orderbook
+// After a transaction is mined we calculate the surplus and fees for each
+// transaction and insert them into the database (settlement_observations
+// table).
+
+use {
+    crate::{
+        database::{
+            on_settlement_event_updater::{AuctionData, SettlementUpdate},
+            Postgres,
+        },
+        decoded_settlement::DecodedSettlement,
+        domain::{eth, settlement::Transaction},
+        infra,
+    },
+    anyhow::{Context, Result},
+    async_trait::async_trait,
+    database::PgTransaction,
+    primitive_types::H256,
+    shared::external_prices::ExternalPrices,
+    sqlx::PgConnection,
+    std::{
+        collections::{BTreeMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::{Mutex, Notify},
+};
+
+pub struct OnSettlementEvent {
+    inner: Arc<Inner>,
+}
+
+struct Inner {
+    notify: Notify,
+}
+
+enum AuctionIdRecoveryStatus {
+    /// The auction id was recovered and the auction data should be added.
+    AddAuctionData(i64, DecodedSettlement),
+    /// The auction id was recovered but the auction data should not be added.
+    DoNotAddAuctionData(i64),
+    /// The auction id was not recovered.
+    InvalidCalldata,
+}
+
+pub struct CircuitBreaker {
+    circuit_breaker: infra::blockchain::circuit_breaker::CircuitBreaker,
+    db: Postgres,
```

Comment:
remove db depencency. If we need something from db we will get it through `infra::Persistence`